src/Orleans/Utils/IKeyedServiceDescriptor.cs

```diff
@@ -0,0 +1,85 @@
+﻿
+using System;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans
+{
+    public static class KeyedServiceDescriptorExtensions
+    {
+        /// <summary>
+        /// Add supported for keyed services to a service collection
+        /// </summary>
+        public static IServiceCollection SupportKeyedServices(this IServiceCollection collection)
+        {
+            return collection.AddSingleton(typeof(IKeyedServiceCollection<,>), typeof(KeyedServiceCollection<,>));
+        }
+
+        /// <summary>
+        /// Associates a key with a specialization of a service type.
+        /// </summary>
+        public static IServiceCollection WithKey<TKey, TService, TSpecialization>(this IServiceCollection collection, TKey key)
+            where TService : class
+            where TSpecialization : class, TService
+        {
+            return collection.AddSingleton<IKeyedServiceDescriptor<TKey, TService>>(sp => new KeyedServiceDescriptor<TKey, TService, TSpecialization>(key));
```

Comment:
yes, but this is just the descriptor, their is no need for it to be anything other than singleton, because it only maps the key to the specific type.